modules/index-rebuild/src/main/java/org/opencastproject/index/rebuild/IndexRebuildService.java

```diff
@@ -0,0 +1,232 @@
+    }
+  }
+
+  /**
+   * Partially rebuild the index from a specific service.
+   *
+   * @param index
+   *           The index to rebuild.
+   * @param serviceName
+   *           The name of the {@link Service}.
+   *
+   * @throws IllegalArgumentException
+   *           Thrown if the service doesn't exist.
+   * @throws IndexRebuildException
+   *           Thrown if the index rebuild failed.
+   */
+  public synchronized void rebuildIndex(AbstractSearchIndex index, String serviceName)
+          throws IllegalArgumentException, IndexRebuildException {
+    IndexRebuildService.Service service = IndexRebuildService.Service.valueOf(serviceName);
+    logger.info("Starting partial rebuild of index '{}' from service '{}'.", index.getIndexName(), service);
+    rebuildIndex(index, service);
+  }
+
+  /**
+   * Trigger repopulation of the index with data from a specific service.
+   *
+   * @param index
+   *           The index to rebuild.
+   * @param service
+   *          The {@link IndexRebuildService.Service} to re-add data from.
+   *
+   * @throws IndexRebuildException
+   *           Thrown if the index rebuild failed.
+   */
+  private void rebuildIndex(AbstractSearchIndex index, IndexRebuildService.Service service)
+          throws IndexRebuildException {
+
+    if (!indexProducers.containsKey(service)) {
+      throw new IllegalStateException(format("Service %s is not available", service));
+    }
+
+    IndexProducer indexProducer = indexProducers.get(service);
+    logger.info("Starting to rebuild index '{}' from service '{}'", index.getIndexName(), service);
+    indexProducer.repopulate(index.getIndexName());
+    logger.info("Finished to rebuild index '{}' from service '{}'", index.getIndexName(), service);
+  }
+
+  /**
+   * Add IndexProducer to internal map.
+   *
+   * @param indexProducer
+   *           The IndexProducer to add.
+   */
+  private void addIndexProducer(IndexProducer indexProducer) {
+    if (indexProducers.putIfAbsent(indexProducer.getService(), indexProducer) == null) {
+      logger.info("Service {} added.", indexProducer.getService());
+    }
+  }
+
+  /**
+   * Remove IndexProducer from internal map.
+   *
+   * @param indexProducer
+   *           The IndexProducer to remove.
+   */
+  private void removeIndexProducer(IndexProducer indexProducer) {
+    if (indexProducers.remove(indexProducer.getService(), indexProducer)) {
+      logger.info("Service {} removed.", indexProducer.getService());
+    }
+  }
+
+  /**
+   * Unregister this service at OSGI.
+   */
+  private void unregisterIndexRebuildService() {
+    if (serviceRegistration != null)  {
+      logger.info("Unregister IndexRebuildService.");
+      serviceRegistration.unregister();
+      serviceRegistration = null;
+    }
+  }
+
+  /**
+   * Register this service at OSGI.
+   *
+   * @param bundleContext
+   *           The bundle context.
+   */
+  private void registerIndexRebuildService(BundleContext bundleContext) {
```

Comment:
I just realized that the registration isn't actually thread-safe. I think it would be if you just mark this method as synchronized.

Reply:
Dammit, multi-threading was never a strong area of mine... Fixed.